src/cargo/util/lock.rs

```diff
@@ -0,0 +1,92 @@
+use file_lock::filename::{Mode, ParseError};
+use file_lock::filename::Lock as FileLock;
+use file_lock::filename::Error as FileLockError;
+
+use std::path::PathBuf;
+use std::fs;
+use std::thread::sleep_ms;
+
+use util::{Config, CargoError, CargoResult, human, caused_human};
+
+pub use file_lock::filename::Kind as LockKind;
+
+impl From<FileLockError> for Box<CargoError> {
+    fn from(t: FileLockError) -> Self {
+        Box::new(t)
+    }
+}
+
+impl From<ParseError> for Box<CargoError> {
+    fn from(t: ParseError) -> Self {
+        Box::new(t)
+    }
+}
+
+impl CargoError for FileLockError {
+    fn is_human(&self) -> bool { true }
+}
+impl CargoError for ParseError {}
+
+pub struct CargoLock {
+    kind: LockKind, 
+    inner: FileLock,
+}
+
+impl CargoLock {
+
+    pub fn lock_kind(config: &Config) -> CargoResult<LockKind> {
+        // TODO(ST): rename config key to something more suitable
+        const CONFIG_KEY: &'static str = "build.lock-kind";
+        match try!(config.get_string(CONFIG_KEY)).map(|t| t.0) {
+            None => Ok(LockKind::NonBlocking),
+            Some(kind_string) => match kind_string.parse() {
```

Comment:
In this case the `FromStr` implementation is located in the `file_lock` crate, and I would expect the actual parsing of a value to happen in this crate. If we're just gonna go with "always wait" though this isn't so bad.